### PR TITLE
fix(core): fix lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case

### PR DESCRIPTION
**Pull request by oowl**
### Summary

backport patch in https://github.com/openresty/lua-nginx-module/pull/2323

Context: https://github.com/openresty/lua-nginx-module/issues/2320

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

FTI-6005

### CODE REVIEW

File: changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+message: fix lua-nginx-module context was clear when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```
Comment (chronolaw) [inline changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml:1]: Should be "fixed an issue where ..."

Comment (oowl) [reply changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml]: done

File: changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+message: fix an issues where lua-nginx-module context was clear when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```
Comment (chronolaw) [inline changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml:1]: Should be `was cleared`?

Comment (oowl) [reply changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml]: done

File: build/openresty/patches/nginx-1.25.3_07-fix-lua-context-clean-by-send-header.patch [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+             dd("sending headers");
+             rc = ngx_http_send_header(r);
++            if (r->filter_finalize) {
++                ngx_http_set_ctx(r, ctx, ngx_http_lua_module);
++            }
++
+             ctx->header_sent = 1;
+             return rc;
+         }
+diff --git a/bundle/ngx_lua-0.10.26/t/002-content.t b/bundle/ngx_lua-0.10.26/t/002-content.t
```
Comment (chronolaw) [inline build/openresty/patches/nginx-1.25.3_07-fix-lua-context-clean-by-send-header.patch:16]: We do not run openresty's test in kong, should we remove this part in patch?

Comment (chronolaw) [reply build/openresty/patches/nginx-1.25.3_07-fix-lua-context-clean-by-send-header.patch:16]: See: https://github.com/Kong/kong/pull/12940#discussion_r1583400706

File: changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+message: fix an issues where lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```
Comment (chronolaw) [inline changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml:1]:
```suggestion
message: fix an issue where lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```

File: build/openresty/patches/nginx-1.25.3_07-fix-lua-context-clean-by-send-header.patch (hunk 3 above)
Comment (oowl) [reply build/openresty/patches/nginx-1.25.3_07-fix-lua-context-clean-by-send-header.patch:16]: It does not matter if we remove or not remove it. But I think we must keep these changes in our patch to keep our repo sync with openresty upstream.

File: changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+message: fix an issue where lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```
Comment (windmgc) [inline changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml:1]:
```suggestion
message: Fixed an issue where lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```

**Review by windmgc — APPROVED** [review]

**Review by chronolaw — APPROVED** [review]

**Comment by oowl** [comment]
hold on, I need to fix a test case conflict issue, So please do not do any rebase until I fix all the CI.

File: changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+message: Fixed an issue where lua-nginx-module context was cleared when ngx.send_header() trigger filter_finalize case [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323)
```
Comment (outsinre) [inline changelog/unreleased/kong/fix-filter-finalize-in-send-header-clear-context.yml:1]:
```suggestion
message: Fixed an issue where `lua-nginx-module` context was cleared when `ngx.send_header()` triggered `filter_finalize` [openresty/lua-nginx-module#2323](https://github.com/openresty/lua-nginx-module/pull/2323).
```

**Review by outsinre — APPROVED** [review]

**Review by chronolaw — APPROVED** [review]

**Review by outsinre — APPROVED** [review]

**Pull request merged**